### ISSUE TITLE
Further improvements to HTML code viewer

### ISSUE THREAD

**Issue by j-f1** [denoland/registry#102]
- [ ] Support highlighting column numbers (for error messages; perhaps underline up to the next word boundary) (from https://github.com/denoland/registry/pull/101#discussion_r295798967)
- [ ] "copy permalink" - if user is viewing master branch there should be option to copy URL with current commit hash (from https://github.com/denoland/registry/pull/101#issuecomment-504028298)

**Comment by MarkTiedemann** [comment]
I think the [Monaco Editor](https://microsoft.github.io/monaco-editor/index.html) could be used for making types clickable - it has "Go to Symbol..." and "Go to references...". This would also improve syntax highlighting. (However, it's not supported on mobile.)

**Comment by ry** [comment]
It's going to require a custom parser (using TS) to accomplish the "go to reference"

**Comment by j-f1** [comment]
Here’s my prototype of a “Jump to Definition” feature: https://observablehq.com/d/897d36265e5ef353. Here’s how it works:

- Parse the source using TypeScript’s API
- Walk the AST, searching for identifiers.
- For each identifier:
  - if it’s a declaration (*i.e. the symbol the type checker has created for that variable has the identifier in its list of declarations*), replace it with `<span class="definition" id="foo-42">foo</span>` (42 = the start index of the variable)
  - otherwise (*i.e. it’s a usage*), replace it with `<a class="ref" href="#foo-42">foo</a>`

I use a special unescaping mechanism to wrap the HTML (`foo` → ``<#'"`*/foo/*`"'#>``) which is a pattern that is invalid syntax in every JS context I can think of. This allows the renderer to escape the HTML, then go back and unescape the HTML code that needs to be embedded.

**Comment by ry** [comment]
Your observable notebook is awesome on many levels. Very cool.

I feel like the URL to [this function definition](https://deno.land/std@14fefcb/http/server.ts#L18) should be https://deno.land/std@14fefcb/http/server.ts#bufWriter 
So can you omit "the start index of the variable" and just leave the variable?

> I use a special unescaping mechanism

I don't understand. Sounds sketch.

**Comment by ry** [comment]
Any idea how we can add some unit tests to your code? I'm feeling like this is getting complex enough that we want some tests.

**Comment by j-f1** [comment]
> Your observable notebook is awesome on many levels ^_^

Thank you ❤️

> So can you omit "the start index of the variable" and just leave the variable?

Unfortunately I can’t do this because there can be duplicates — variables can be shadowed, and TS also supports jump-to-definition for property names.

> > I use a special unescaping mechanism
>
> I don't understand. Sounds sketch.

Basically, I need to embed HTML at specific locations in the source code while also escaping any special characters in the source code. I’ll try to roll my own function that does escaping and embedding simultaneously to avoid needing this.

> Any idea how we can add some unit tests to your code? I'm feeling like this is getting complex enough that we want some tests.

Sure, and I’ll definitely do that when I write the actual code. The notebook was more of a sketch/exploration of how the implementation would work.

**Comment by j-f1** [comment]
Update: I rewrote the embedding code to avoid the sketchy escaping mechanism and added support for clicking on imports (this will be done server-side rather than client-side, and it’ll use the TS parser rather than relying on reading CodeMirror’s HTML). However, there’s a small bug where overlapping escapes cause undesired results. Because of that, I’m planning to rewrite it again.

**Comment by ry** [comment]
@j-f1 

> I rewrote the embedding code to avoid the sketchy escaping mechanism

Rad.

> > So can you omit "the start index of the variable" and just leave the variable?
> Unfortunately I can’t do this because there can be duplicates 

That makes sense. But what about for top-level exports of a module? There will ever only be one "fooBar" export. It would be very nice to have canonical URLs for these symbols: https://example.com/script.ts#fooBar

Does your program work with TS only or also JS?

**Comment by j-f1** [comment]
> But what about for top-level exports of a module?

Good point — I’ll look at that next.

> Does your program work with TS only or also JS?

It works with JS too since JS is a subset of TS! There isn’t as much type info though, so linking to type definitions for object properties doesn’t work.

**Comment by j-f1** [comment]
Just confirming @ry — for auto-generated docs, `typedoc` should be used?

**Comment by ry** [comment]
@j-f1 No, typedoc should not be used. Only typescript should be used - no other dependencies.

@bartlomieju is also thinking about this problem for "deno doc" the command-line version.

**Comment by bartlomieju** [comment]
Just FYI I intend to pull out JSDoc for all export from a file using TypeScript AST, but I'll start working on it in the second part of the week. @j-f1 in the meantime feel free to hack around on your own

**Comment by ry** [comment]
Maybe "deno doc" can output a JSON form of the docs, which the registry can call and consume.

**Comment by j-f1** [comment]
Or that could be a package in the standard library, with a CLI that can be called to generate docs for a given file/project and open `deno.land`-style rendered pages in the browser.
